### PR TITLE
feat(utxo-lib): Allow decimal fee rates

### PR DESCRIPTION
**Pull request by dahaca**
Allow decimal fee rates for #3252

### CODE REVIEW

File: packages/utxo-lib/src/coinselect/utils.ts [hunk 1]
```diff
@@ -116,11 +116,19 @@ export function dustThreshold(feeRate: number, options: CoinSelectOptions) {
     return Math.max(options.dustThreshold, price);
 }
 
-export function uintOrNaN(v: number) {
-    if (typeof v !== 'number') return;
-    if (Number.isNaN(v) || !Number.isFinite(v) || !Number.isInteger(v) || v < 0) return;
-    return v;
-}
+export const formatStringFeeRate = (feeRate: string): number | undefined => {
```
Comment (szymonlesisz) [inline packages/utxo-lib/src/coinselect/utils.ts]:
```suggestion
export const formatStringFeeRate = (feeRate: string) => {
```
nit: you dont have to define return type, typescript will do it for you

File: packages/utxo-lib/tests/coinselect/utils.test.ts [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
+        // @ts-expect-error invalid arg
+        expect(formatStringFeeRate(NaN)).toBeUndefined();
+        // @ts-expect-error invalid arg
+        expect(formatStringFeeRate(Infinity)).toBeUndefined();
     });
 
     it('bignumberOrNaN', () => {
         expect(bignumberOrNaN('1')).not.toBeUndefined();
-        expect(bignumberOrNaN('1.1')).toBeUndefined();
+        // expect(bignumberOrNaN('1.1')).toBeUndefined();
```
Comment (szymonlesisz) [inline packages/utxo-lib/tests/coinselect/utils.test.ts]: why commented? those should be still valid (invalid) cases, no?

File: packages/utxo-lib/tests/__fixtures__/compose.ts [hunk 3]
```diff
@@ -344,6 +344,77 @@ export default [
             type: 'error',
         },
     },
+    {
```
Comment (szymonlesisz) [inline packages/utxo-lib/tests/__fixtures__/compose.ts:405]: please also add at least one valid `feeRate with decimal value` test case

**Review by szymonlesisz — CHANGES_REQUESTED** [review]
since you are doing changes in two different packages, 1 public and 1 private i would split it in to two commits for better transparency:
- `fix(utxo-lib): allow decimal values in feeRate`
- `fix(suite): ....`

File: packages/utxo-lib/src/coinselect/utils.ts (hunk 1 above)
Comment (dahaca) [reply packages/utxo-lib/src/coinselect/utils.ts]: Sure, it's just another guard, in my view makes sense for utils to have types defined as thoroughly as possible, so that TS throws if a return doesn't match the type after some edits :)

Comment (szymonlesisz) [reply packages/utxo-lib/src/coinselect/utils.ts]: right, but if you mess something here and for example return a string, then TS will catch it anyway in a place where you are using this utility (compose/index)

File: packages/utxo-lib/src/compose/index.ts [hunk 4]
```diff
@@ -4,6 +4,7 @@ import * as transaction from './transaction';
 import { coinselect } from './coinselect';
 
 import type { Result } from './coinselect';
+import { formatStringFeeRate } from '../coinselect/utils';
```
Comment (szymonlesisz) [inline packages/utxo-lib/src/compose/index.ts]: pls, move this import to line 5 (before import type) or even better, join imports from `./coinselect` (line 4 + line 6) - looks ilke me and the reviewer didn't noticed that while moving this lib from the origin :)

File: packages/utxo-lib/src/coinselect/utils.ts (hunk 1 above)
Comment (szymonlesisz) [reply packages/utxo-lib/src/coinselect/utils.ts]: but it doesn't matter really, it's only my personal preference, i would not define it.
i'm leaving this decision to you

File: packages/utxo-lib/src/compose/index.ts (hunk 4 above)
Comment (dahaca) [reply packages/utxo-lib/src/compose/index.ts]: done 463ab037d9c46a43792a5ad5b0e6f993c4cf7d1d

File: packages/utxo-lib/tests/__fixtures__/compose.ts (hunk 3 above)
Comment (dahaca) [reply packages/utxo-lib/tests/__fixtures__/compose.ts:405]: Added but not sure about the data in `transaction` 463ab037d9c46a43792a5ad5b0e6f993c4cf7d1d

File: packages/utxo-lib/tests/coinselect/utils.test.ts (hunk 2 above)
Comment (dahaca) [reply packages/utxo-lib/tests/coinselect/utils.test.ts]: A typo somehow... Fixed 463ab037d9c46a43792a5ad5b0e6f993c4cf7d1d

File: packages/utxo-lib/src/compose/index.ts (hunk 4 above)
Comment (szymonlesisz) [reply packages/utxo-lib/src/compose/index.ts]: almost :)
should be:
```
import { coinselect, Result } from './coinselect';
```
it's not about the order, i never saw `import { type .. }` syntax before and i'm quite surprised that TS is accepting that? is this something new?

Comment (dahaca) [reply packages/utxo-lib/src/compose/index.ts]: Yep, it was [introduced in TS 4.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#type-modifiers-on-import-names) :) Syntax sugar to avoid having 2 lines of imports.

File: packages/utxo-lib/src/coinselect/utils.ts (hunk 1 above)
Comment (dahaca) [reply packages/utxo-lib/src/coinselect/utils.ts]: Ok, removed affdf675ab7412f0fd13ceaa5ec53ce84fcd708c

File: packages/utxo-lib/src/compose/index.ts (hunk 4 above)
Comment (dahaca) [reply packages/utxo-lib/src/compose/index.ts]: Or do you want me to just get rid of the `type`?

Comment (szymonlesisz) [reply packages/utxo-lib/src/compose/index.ts]: i rather do, i've looked at code and we are not using it anywhere, lets be consistent

Comment (dahaca) [reply packages/utxo-lib/src/compose/index.ts]: 226eda709e01d863dee3d9d87b8a9fd608a0e01e

File: packages/utxo-lib/tests/compose.test.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import { composeTx } from '../src';
+import * as utils from '../src/compose/utils';
```
Comment (dahaca) [inline packages/utxo-lib/tests/compose.test.ts:2]: Why not just

```
import { convertFeeRate } from '../src/compose/utils';
```

File: packages/utxo-lib/src/compose/index.ts [hunk 6]
```diff
@@ -47,14 +47,14 @@ export function composeTx({
 
     const splitOutputs = request.splitByCompleteness(outputs);
 
-    let csResult: Result = { type: 'false' };
+    let csResult: ReturnType<typeof coinselect> = { type: 'false' };
```
Comment (dahaca) [inline packages/utxo-lib/src/compose/index.ts:50]: Why this change? Also, `type Result` in `coinselect` is now unused

File: packages/utxo-lib/tests/compose.test.ts (hunk 5 above)
Comment (szymonlesisz) [reply packages/utxo-lib/tests/compose.test.ts:2]: this tests should be extended in the future to cover all the utilities in that file, not only this one

File: packages/utxo-lib/src/compose/index.ts (hunk 6 above)
Comment (szymonlesisz) [reply packages/utxo-lib/src/compose/index.ts:50]: because it's cleaner :) and type is still used (compose/coinselect line 44)

File: packages/utxo-lib/tests/compose.test.ts (hunk 5 above)
Comment (dahaca) [reply packages/utxo-lib/tests/compose.test.ts:2]: makes sense. I'd still destructure them below imports to not have a ton of `utils.` :D Up to you.

File: packages/utxo-lib/src/compose/index.ts (hunk 6 above)
Comment (dahaca) [reply packages/utxo-lib/src/compose/index.ts:50]: True but the `export` is meaningless. That's a nitpick though

Comment (szymonlesisz) [reply packages/utxo-lib/src/compose/index.ts:50]: feel free to remove the export then

File: packages/utxo-lib/tests/compose.test.ts (hunk 5 above)
Comment (szymonlesisz) [reply packages/utxo-lib/tests/compose.test.ts:2]: leaving as is, it's only a test

**Review by szymonlesisz — APPROVED** [review]
